File: src/core/contracts/dto.ts
/**
 * IPC DTOs — the data shapes crossing the webview <-> Rust-core seam.
 *
 * These MUST stay in sync with `src-tauri/src/dto.rs` (serde structs).
 * This file is the single source of truth for the TS side of the contract.
 * Changing a shape here is a contract change — version it deliberately.
 *
 * See architecture.md §3 "IPC-контракт".
 */

/**
 * Region of the screen to capture. Absent = full primary display.
 *
 * Coordinates are PHYSICAL device pixels (matching the captured frame), NOT
 * logical/CSS pixels. A future region-select UI must convert from CSS pixels
 * using the display scale factor before sending, or the crop will be offset on
 * scaled (125%/150%) displays. See `scap_capturer.rs` `bgra_to_rgba` DPI caveat.
 */
export interface CaptureRegionDto {
  readonly x: number;
  readonly y: number;
  readonly width: number;
  readonly height: number;
}

export interface CaptureRequestDto {
  /** Optional sub-region; omit for full screen. */
  readonly region?: CaptureRegionDto;
  /** Target display index; omit for primary. */
  readonly displayIndex?: number;
}

export interface CaptureResultDto {
  /** PNG bytes, base64-encoded (kept simple for the IPC boundary). */
  readonly imageBase64: string;
  readonly width: number;
  readonly height: number;
  /** Unix epoch milliseconds, UTC. */
  readonly capturedAt: number;
}

export interface OcrRequestDto {
  readonly imageBase64: string;
  /** BCP-47 hints, e.g. ["en", "ru"]. */
  readonly languages?: readonly string[];
}

export interface OcrBoxDto {
  readonly text: string;
  readonly confidence: number;
  readonly box: CaptureRegionDto;
}

export interface OcrResultDto {
  readonly fullText: string;
  readonly boxes: readonly OcrBoxDto[];
}

// --- LLM streaming (secure-native, `llm_stream`) ---------------------------
// The webview assembles the request (model slug via ResilientLlm + messages) and
// hands it to native, which owns the key and the HTTP call. Chunks come back
// over a Tauri Channel<LlmChunkDto>. Mirror of the LlmPort shapes, kept as DTOs
// because they cross the seam. See architecture.md §3, §6.

export interface LlmTextPartDto {
  readonly kind: 'text';
  readonly text: string;
}
export interface LlmImagePartDto {
  readonly kind: 'image';
  /** base64 PNG. */
  readonly imageBase64: string;
}
export type LlmContentPartDto = LlmTextPartDto | LlmImagePartDto;

export interface LlmMessageDto {
  readonly role: 'system' | 'user' | 'assistant';
  readonly parts: readonly LlmContentPartDto[];
}

export interface LlmStreamRequestDto {
  /** OpenRouter model slug chosen by the resilient LLM layer (webview side). */
  readonly model: string;
  readonly messages: readonly LlmMessageDto[];
  /** Sampling temperature; omitted = not sent to the provider (R16). */
  readonly temperature?: number;
  /** OpenRouter `reasoning.effort` ('low'|'medium'|'high'); omitted = no reasoning (R16). */
  readonly reasoningEffort?: string;
  /** Longest image edge in px; native downscales larger images before sending (R4/R17). */
  readonly maxImageEdge?: number;
  /** Client-generated id so `llm_cancel` can stop this request. */
  readonly requestId?: string;
}

export interface LlmCancelRequestDto {
  readonly requestId: string;
}

export interface LlmUsageDto {
  readonly inputTokens: number;
  readonly outputTokens: number;
  /** USD cost reported by OpenRouter (`usage.cost`), when present. */
  readonly cost?: number;
}

/** Discriminated union streamed over the Channel; terminal chunk is finish|error. */
export type LlmChunkDto =
  | { readonly type: 'text-delta'; readonly delta: string }
  | {
      readonly type: 'finish';
      readonly reason: string;
      readonly usage?: LlmUsageDto;
      /** Model slug the provider reports as having served the request (R19). */
      readonly model?: string;
    }
  // `retryable` tells the webview whether to fail over to the next model in the
  // chain (transient/provider-side error) or surface the failure as-is (missing
  // key, auth, bad request). Mirrors `LlmChunk::Error.retryable` in dto.rs.
  | { readonly type: 'error'; readonly message: string; readonly retryable: boolean };

export interface SecretGetRequestDto {
  readonly key: string;
}
export interface SecretGetResultDto {
  readonly value: string | null;
}
export interface SecretSetRequestDto {
  readonly key: string;
  readonly value: string;
}

// --- Audio → STT (secure-native, phase 9) ----------------------------------
// `audioStartCapture`/`audioStopCapture` take no payload. `transcribeAudio`
// returns this. The recorded audio stays in native; only the text crosses the
// seam. Mirror of `TranscribeResult` in dto.rs.
export interface TranscribeResultDto {
  /** Recognized transcript (may be empty if nothing was captured). */
  readonly text: string;
}

File: src-tauri/src/dto.rs
//! IPC DTOs — the Rust side of the webview <-> core contract seam.
//! MUST stay in sync with `src/core/contracts/dto.ts`. `camelCase` matches TS.

use serde::{Deserialize, Serialize};

/// Coordinates are PHYSICAL device pixels (matching the captured frame), not
/// logical/CSS pixels — see the DPI caveat in `infra/scap_capturer.rs`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureRegion {
    pub x: u32,
    pub y: u32,
    pub width: u32,
    pub height: u32,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureRequest {
    #[serde(default)]
    pub region: Option<CaptureRegion>,
    #[serde(default)]
    pub display_index: Option<u32>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CaptureResult {
    pub image_base64: String,
    pub width: u32,
    pub height: u32,
    /// Unix epoch milliseconds, UTC.
    pub captured_at: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OcrRequest {
    pub image_base64: String,
    #[serde(default)]
    pub languages: Option<Vec<String>>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OcrBox {
    pub text: String,
    pub confidence: f32,
    // `box` is a Rust keyword, so the field is named `box_region` on this
    // side but serialized as `box` to match `OcrBoxDto.box` in dto.ts —
    // the wire shape is the single source of truth (architecture.md §3).
    #[serde(rename = "box")]
    pub box_region: CaptureRegion,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct OcrResult {
    pub full_text: String,
    pub boxes: Vec<OcrBox>,
}

// --- LLM streaming (secure-native, `llm_stream`) ---------------------------
// Mirror of `core/contracts/dto.ts` LLM DTOs (phase 2). Frozen here ahead of
// the phase-2 command so the seam can't drift while the Rust side is unused.

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum LlmRole {
    System,
    User,
    Assistant,
}

// `rename_all` renames the variant *tags* (Text -> "text", Image -> "image");
// `rename_all_fields` is what renames the *inner* fields (image_base64 ->
// "imageBase64"). Without the latter, the `Image` variant deserializes its
// field as snake_case `image_base64` and rejects the webview's camelCase
// `imageBase64`, failing with `missing field \`image_base64\``. The wire shape
// in `dto.ts` (LlmImagePartDto.imageBase64) is the source of truth.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "kind", rename_all = "camelCase", rename_all_fields = "camelCase")]
pub enum LlmContentPart {
    Text { text: String },
    Image { image_base64: String },
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LlmMessage {
    pub role: LlmRole,
    pub parts: Vec<LlmContentPart>,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LlmStreamRequest {
    /// OpenRouter model slug chosen by the resilient LLM layer (webview side).
    pub model: String,
    pub messages: Vec<LlmMessage>,
    /// Sampling temperature; `None` = not sent (provider default). Per route,
    /// chosen by the webview (R16). `f64` so `0.3` serializes as `0.3`.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub temperature: Option<f64>,
    /// OpenRouter `reasoning.effort` ("low" | "medium" | "high"); `None` = no
    /// reasoning requested (R16).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reasoning_effort: Option<String>,
    /// Longest image edge in px; larger images are area-downscaled before the
    /// request (R4/R17). `None` = send as captured.
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub max_image_edge: Option<u32>,
    /// Client-generated id for `llm_cancel` (stop / superseded run). `None` =
    /// not cancellable (older webviews, tests).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub request_id: Option<String>,
}

/// Payload of `llm_cancel`.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LlmCancelRequest {
    pub request_id: String,
}

#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct LlmUsage {
    pub input_tokens: u32,
    pub output_tokens: u32,
    /// USD cost reported by OpenRouter (`usage.cost`), when present (R9).
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub cost: Option<f64>,
}

/// Discriminated union streamed over `Channel<LlmChunk>`; terminal chunk is
/// `finish` or `error`. Mirror of `LlmChunkDto` in dto.ts.
#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "kebab-case")]
pub enum LlmChunk {
    TextDelta {
        delta: String,
    },
    Finish {
        reason: String,
        #[serde(default, skip_serializing_if = "Option::is_none")]
        usage: Option<LlmUsage>,
        /// Model slug the provider reports as having served the request (R19).
        #[serde(default, skip_serializing_if = "Option::is_none")]
        model: Option<String>,
    },
    Error {
        message: String,
        /// Whether this failure is worth retrying on a DIFFERENT model. The
        /// webview's resilient LLM layer uses it to decide whether to fail over
        /// to the next model in the chain. Mirrors `LlmError.retryable`
        /// (llm.port.ts) and the `LlmChunkDto` error variant (dto.ts).
        retryable: bool,
    },
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretGetRequest {
    pub key: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretGetResult {
    pub value: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SecretSetRequest {
    pub key: String,
    pub value: String,
}

// --- Audio → STT (secure-native, phase 9) ----------------------------------
// The webview toggles recording via `audio_start_capture`/`audio_stop_capture`
// (no payload) and, at send time, calls `transcribe_audio` which returns this.
// The recorded audio itself never crosses the seam — only the text. Mirror of
// `TranscribeResultDto` in dto.ts.

/// Result of `audio_start_capture`: the rolling window the recording keeps
/// (the LAST `max_seconds`), so the HUD can tell the user.
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AudioStartResult {
    pub max_seconds: u32,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TranscribeResult {
    /// Recognized transcript (may be empty if nothing was captured).
    pub text: String,
}

#[cfg(test)]
mod tests {
    use super::*;

    // Locks the webview <-> Rust wire shape for the image content part. The
    // webview (dto.ts: LlmImagePartDto) sends camelCase `imageBase64`; a serde
    // config renaming only the variant tag (not the inner field) regressed this
    // to snake_case `image_base64` and broke `llm_stream` with an "invalid args
    // `request` ... missing field `image_base64`" rejection.
    #[test]
    fn deserializes_camel_case_image_content_part_from_the_webview() {
        let json = r#"{
            "model": "anthropic/claude-haiku-4.5",
            "messages": [
                { "role": "user", "parts": [
                    { "kind": "text", "text": "describe" },
                    { "kind": "image", "imageBase64": "QUJD" }
                ] }
            ]
        }"#;

        let request: LlmStreamRequest = serde_json::from_str(json).unwrap();

        assert_eq!(request.model, "anthropic/claude-haiku-4.5");
        assert_eq!(request.messages[0].role, LlmRole::User);
        assert_eq!(
            request.messages[0].parts,
            vec![
                LlmContentPart::Text {
                    text: "describe".to_string()
                },
                LlmContentPart::Image {
                    image_base64: "QUJD".to_string()
                },
            ]
        );
    }

    #[test]
    fn serializes_image_content_part_back_to_camel_case() {
        let part = LlmContentPart::Image {
            image_base64: "QUJD".to_string(),
        };
        let value = serde_json::to_value(&part).unwrap();
        assert_eq!(value["kind"], "image");
        assert_eq!(value["imageBase64"], "QUJD");
    }

    #[test]
    fn request_profile_fields_are_optional_and_camel_case() {
        // Old webviews (no profile fields) must still deserialize.
        let bare: LlmStreamRequest =
            serde_json::from_str(r#"{ "model": "m", "messages": [] }"#).unwrap();
        assert_eq!(bare.temperature, None);
        assert_eq!(bare.reasoning_effort, None);
        assert_eq!(bare.max_image_edge, None);

        let full: LlmStreamRequest = serde_json::from_str(
            r#"{ "model": "m", "messages": [], "temperature": 0.3,
                 "reasoningEffort": "medium", "maxImageEdge": 2576 }"#,
        )
        .unwrap();
        assert_eq!(full.temperature, Some(0.3));
        assert_eq!(full.reasoning_effort.as_deref(), Some("medium"));
        assert_eq!(full.max_image_edge, Some(2576));
    }

    #[test]
    fn finish_chunk_serializes_model_and_cost_in_camel_case() {
        let chunk = LlmChunk::Finish {
            reason: "stop".to_string(),
            usage: Some(LlmUsage {
                input_tokens: 1,
                output_tokens: 2,
                cost: Some(0.5),
            }),
            model: Some("anthropic/claude-haiku-4.5".to_string()),
        };
        let value = serde_json::to_value(&chunk).unwrap();
        assert_eq!(value["type"], "finish");
        assert_eq!(value["usage"]["inputTokens"], 1);
        assert_eq!(value["usage"]["cost"], 0.5);
        assert_eq!(value["model"], "anthropic/claude-haiku-4.5");
    }
}
